File: yetbota-web/src/types/moderation.ts
// Moderation & reporting — see moderation-frontend-integration.md.
// Admin dashboard talks to the content-service REST routes under
// /v1/admin/moderation/* (envelope-unwrapped by contentBaseApi).

export type ModContentType = "POST" | "COMMENT";
export type ReportReason = "SPAM" | "OFFENSIVE" | "INCORRECT" | "OTHER";
export type CaseStatus = "PENDING" | "RESOLVED" | "REJECTED";
export type ModerationStatus = "VISIBLE" | "HIDDEN" | "REMOVED";
export type AdminAction = "DELETE" | "DISMISS" | "BAN";
export type CaseResolution = "DELETED" | "DISMISSED";

export interface ModerationCase {
  id: string;
  content_type: ModContentType;
  content_id: string;
  report_count: number;
  status: CaseStatus;
  // Currently equals report_count — use for sort/priority.
  severity: number;
  auto_hidden: boolean;
  first_reported_at: string;
  last_reported_at: string;
  // Optimistic-concurrency token; echo it back when acting on the case.
  version: number;
  // Present once the case is resolved/rejected.
  resolved_by?: string;
  resolved_at?: string;
  resolution?: CaseResolution;
}

export interface CasePreview {
  content_type: ModContentType;
  content_id: string;
  author_id: string;
  title?: string;
  snippet: string;
  moderation_status: ModerationStatus;
  // true when the underlying content row is gone — render a placeholder.
  missing: boolean;
}

export interface ModerationCaseListItem {
  case: ModerationCase;
  preview: CasePreview;
}

export interface ListCasesQuery {
  status?: CaseStatus;
  reason?: ReportReason;
  content_type?: ModContentType;
  page?: number;
  page_size?: number;
}

export interface ListCasesData {
  cases: ModerationCaseListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface ModerationReport {
  id: string;
  reporter_id: string;
  reason: ReportReason;
  details: string;
  created_at: string;
}

export interface ModerationContent {
  id: string;
  content_type: ModContentType;
  author_id: string;
  title?: string;
  body: string;
  is_question?: boolean;
  is_answer?: boolean;
  moderation_status: ModerationStatus;
  created_at: string;
}

export interface CaseDetailData {
  case: ModerationCase;
  reports: ModerationReport[];
  // Omitted when the underlying post/comment no longer exists.
  content?: ModerationContent;
}

export interface CaseActionRequest {
  action: AdminAction;
  note?: string;
  ban_reason?: string;
  version: number;
}

export interface CaseActionData {
  case_id: string;
  status: CaseStatus;
  resolution: CaseResolution;
}

// End-user reporting (REST). Exposed for completeness; admin UI doesn't use it.
export interface CreateReportRequest {
  content_type: ModContentType;
  content_id: string;
  reason: ReportReason;
  details?: string;
}

export interface CreateReportData {
  case_id: string;
  report_count: number;
  status: CaseStatus;
  auto_hidden: boolean;
}

File: backend/content-service/internal/domain/moderation/moderation.go
package moderation

import (
	"context"
	"database/sql"
	"time"
)

const (
	ContentTypePost    = "POST"
	ContentTypeComment = "COMMENT"

	ReasonSpam      = "SPAM"
	ReasonOffensive = "OFFENSIVE"
	ReasonIncorrect = "INCORRECT"
	ReasonOther     = "OTHER"

	CaseStatusPending  = "PENDING"
	CaseStatusResolved = "RESOLVED"
	CaseStatusRejected = "REJECTED"

	ResolutionDeleted   = "DELETED"
	ResolutionDismissed = "DISMISSED"

	ActionDelete  = "DELETE"
	ActionDismiss = "DISMISS"
	ActionBan     = "BAN"
	ActionUnhide  = "UNHIDE"
)

type Report struct {
	ID          string
	ContentType string
	ContentID   string
	ReporterID  string
	Reason      string
	Details     string
	CreatedAt   time.Time
}

type ModerationCase struct {
	ID              string
	ContentType     string
	ContentID       string
	ReportCount     int
	Status          string
	Severity        int
	AutoHidden      bool
	FirstReportedAt time.Time
	LastReportedAt  time.Time
	ResolvedBy      string
	ResolvedAt      *time.Time
	Resolution      string
	Version         int
}

type ModerationAction struct {
	ID         string
	CaseID     string
	AdminID    string
	Action     string
	TargetType string
	TargetID   string
	Note       string
	CreatedAt  time.Time
}

type CaseListOptions struct {
	Status      string
	Reason      string
	ContentType string
	Page        int
	PageSize    int
}

type Repository interface {
	CreateReport(ctx context.Context, tx *sql.Tx, r *Report) (bool, error)
	UpsertCaseOnReport(ctx context.Context, tx *sql.Tx, id, contentType, contentID string, reportedAt time.Time) (*ModerationCase, error)
	MarkCaseAutoHidden(ctx context.Context, tx *sql.Tx, caseID string) error
	ListCases(ctx context.Context, opts *CaseListOptions) ([]*ModerationCase, error)
	CountCases(ctx context.Context, opts *CaseListOptions) (int64, error)
	GetCase(ctx context.Context, id string) (*ModerationCase, error)
	ListReports(ctx context.Context, contentType, contentID string) ([]*Report, error)
	ResolveCase(ctx context.Context, tx *sql.Tx, caseID string, expectedVersion int, status, resolution, resolvedBy string) (bool, error)
	CreateAction(ctx context.Context, tx *sql.Tx, a *ModerationAction) error
}

type RateLimiter interface {
	Allow(ctx context.Context, key string, max int, window time.Duration) (bool, error)
}

type UserBanner interface {
	Ban(ctx context.Context, userID, reason, actorID string) error
}
